apollo-router/src/protocols/sse/mod.rs

```diff
@@ -0,0 +1,138 @@
+            completed: false,
+            terminated: false,
+        })
+    }
+}
+
+pub(crate) fn convert_sse_stream(
+    client: impl client::Client,
+    id: String,
+) -> impl Stream<Item = serde_json::Result<ServerMessage>> {
+    client.stream().filter_map(move |msg| match msg {
+        Ok(sse) => match sse {
+            event_parser::Sse::Event(event) => match event.event_type.as_str() {
+                "next" => future::ready(Some(serde_json::from_str(&event.data).map(|s| {
+                    ServerMessage::Next {
+                        id: event.id.unwrap_or_else(|| id.clone()),
+                        payload: s,
+                    }
+                }))),
+                "complete" => future::ready(Some(Ok(ServerMessage::Complete {
+                    id: event.id.unwrap_or_else(|| id.clone()),
+                }))),
+                event_type => future::ready(Some(Ok(ServerMessage::Error {
+                    id: id.to_string(),
+                    payload: ServerError::Error(
+                        graphql::Error::builder()
+                            .message(format!("invalid event type: {event_type} received"))
+                            .extension_code("SSE_INVALID_EVENT_TYPE")
+                            .build(),
+                    ),
+                }))),
+            },
+            event_parser::Sse::Comment(_) => future::ready(None),
+        },
+        Err(err) => {
+            tracing::trace!("cannot consume more message on sse stream: {err:?}");
+            future::ready(Some(Ok(ServerMessage::Error {
+                id: id.to_string(),
+                payload: ServerError::Error(
+                    graphql::Error::builder()
+                        .message(format!("cannot read message from sse: {err:?}"))
+                        .extension_code("SSE_MESSAGE_ERROR")
+                        .build(),
+                ),
+            })))
+        }
+    })
+}
+
+impl<S> Stream for GraphqlSSE<S>
+where
+    S: Stream<Item = serde_json::Result<ServerMessage>>,
+{
+    type Item = graphql::Response;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut().project();
+
+        match Pin::new(&mut this.stream).poll_next(cx) {
+            Poll::Ready(message) => match message {
+                Some(server_message) => match server_message {
+                    Ok(server_message) => {
+                        if let Some(id) = &server_message.id() {
+                            if this.id != id {
+                                tracing::error!("we should not receive data from other subscriptions, closing the stream");
+                                return Poll::Ready(None);
+                            }
+                        }
+                        if let ServerMessage::Ping { .. } = server_message {
+                            // Send pong asynchronously
```

Comment:
Is it a TODO ?

Reply:
I removed this part, since it was not really used